docs/proposals/20230123-azure-service-operator.md

```diff
@@ -0,0 +1,232 @@
+
+#### Caveats
+
+CAPZ interacts with some Azure services that do not represent infrastructure, and thus cannot be represented in ASO. Resource Health, for example, is "reconciled" by CAPZ currently by getting a resource's health status and reflecting that in the corresponding CAPZ resource, but does not create or update any distinct Azure resources. The new SDK could be used to implement this existing functionality without affecting other service interfaces' use of ASO.
+
+Also, use of the `clusterctl move` command will require extra manual steps to move ASO resources as documented here: https://azure.github.io/azure-service-operator/introduction/frequently-asked-questions/#what-is-the-best-practice-for-transferring-aso-resources-from-one-cluster-to-another.
+
+### Security Model
+
+ASO requires credentials to interact with Azure which are documented here: https://azure.github.io/azure-service-operator/introduction/authentication/. Specifically, ASO needs the values of `$AZURE_CLIENT_ID`, `$AZURE_CLIENT_SECRET`, `$AZURE_SUBSCRIPTION_ID`, and `$AZURE_TENANT_ID` for Service Principal authentication. These credentials must be defined in a Kubernetes Secret that either applies to all ASO resources in all namespaces, all ASO resources in a particular namespace, or individual ASO resources referring to the Secret in its `serviceoperator.azure.com/credential-from` annotation. ASO allows different credentials to be used for different resources to enable multitenancy, similar to CAPZ and its AzureClusterIdentity. To mediate between the two mechanisms, CAPZ will ensure a cluster's AzureClusterIdentity or default parameters from environment variables are reflected in a Secret to be used by ASO and ensure each ASO resource it manages is created with the annotation pointing to the Secret.
+
+Consumers of ASO's APIs (such as CAPZ) also need Kubernetes RBAC permissions to interact with ASO resources. For example, this ClusterRole snippet is required by CAPZ to manage resource groups with ASO:
+```yaml
+- apiGroups:
+  - resources.azure.com
+  resources:
+  - resourcegroups
+  - resourcegroups/status
+  verbs:
+  - get
+  - list
+  - watch
+  - create
+  - update
+  - delete
+```
+The above permissions may overlap with ASO resources already installed by a user not intended to be managed by CAPZ.
+
```

Comment:
With ASO introduction: 
- What happens to AzureClusterIdentity? Do we translate it to a secret that is consumed by ASO? 
- Multi tenancy part looks good. Just adding this comment for clarity. I sense that we can specify to ASO what credentials to use to reconcile the infrastructure. This means for a particular workload cluster all its related ASO resources are reconciled by the same identity as that of specified in the corresponding AzureClusterIdentity.

Reply:
Also for workload identity, 
We need to add an annotation on the concerned ASO pods that tries to authenticate and just tell ASO what client and tenant ID to use. So I do not see any significant challenges as of now apart from the usual workload identity prerequiste steps.

Reply:
> What happens to AzureClusterIdentity? Do we translate it to a secret that is consumed by ASO?

Essentially yes, that's what I was trying to say above with:
> To mediate between the two mechanisms, CAPZ will ensure a cluster's AzureClusterIdentity or default parameters from environment variables are reflected in a Secret to be used by ASO and ensure each ASO resource it manages is created with the annotation pointing to the Secret.

I think AzureClusterIdentity will remain useful even after ASO is fully integrated to be able to define credentials at a per-cluster level which isn't something ASO would be able to handle itself.

Reply:
> We need to add an annotation on the concerned ASO pods that tries to authenticate and just tell ASO what client and tenant ID to use. So I do not see any significant challenges as of now apart from the usual workload identity prerequiste steps.

You don't need to do this actually. You put the client/tenantid into a secret that is attached to the resource(s) in question:
```
 AZURE_SUBSCRIPTION_ID: "$AZURE_SUBSCRIPTION_ID"
 AZURE_TENANT_ID: "$AZURE_TENANT_ID"
 AZURE_CLIENT_ID: "$AZURE_CLIENT_ID"
 USE_WORKLOAD_IDENTITY_AUTH: "true"
```

You don't need to do anything to the operator pod.